Avoid using DenseElementsAttr::getRawData

sdasgup3 opened the pull request:
fixes https://github.com/openxla/stablehlo/issues/83

The PR implements  the underlying data storage of the tensor using an `mlir::HeapAsmResourceBlob` object represents a mutable blob of data. This provides two improvements:
1. No need for using `DenseElementsAttr::getRawData()` which cannot be relied  to provide contiguous representation of underlying data in cases like boolean and splats.
2.  An `mlir::HeapAsmResourceBlob` object, being mutable, simplifies how we store the underlying Tensor data.

burmako commented on docs/interpreter.md:
```diff
@@ -5,11 +5,11 @@
-objects are reference-counted to simplify memory management.
+`Tensor`, which stores the shape (an `mlir::ShapedType` object) and the
+underlying data storage (an `mlir::HeapAsmResourceBlob` object) of the tensor.
+The `mlir::HeapAsmResourceBlob` object represents a mutable blob of data laid
```
Consider simplifying since the name of the class was just called out on the previous line. E.g. "The data storage contains a mutable blob of data...".

sdasgup3 replied:
done

burmako commented on stablehlo/reference/Tensor.cpp:
```diff
@@ -52,65 +52,57 @@ int64_t getSizeInBytes(Type type) {
-                                                    attr.getRawData().data());
+Tensor::Tensor(const Tensor &other) {
+  type_ = other.getType();
+  blob_ = HeapAsmResourceBlob::allocateAndCopy(other.getData(), alignof(char));
```
In the current design, `Tensor::Tensor(const Tensor &other)` and `Tensor &Tensor::operator=(const Tensor &other)` don't involve copying memory ([see details](https://github.com/llvm/llvm-project/blob/dfbaf90043841d18f5871e12b10c674b9b3abf99/llvm/include/llvm/ADT/IntrusiveRefCntPtr.h#L174-L196)), but it looks like in the proposed design, they both involve copying. Do we consider that to be a downside? If yes, is there something that can be done about it?

sdasgup3 replied:
Thanks for catching this. This allows lots of unnecessary copies. Fixed it by reference counting the internal storage.

burmako reviewed (changes requested):
Here are some thoughts on this PR. I'll leave commenting on using `HeapAsmResourceBlob` to Mehdi, since I haven't been following that very closely.

sdasgup3 commented on stablehlo/reference/Tensor.cpp:
```diff
@@ -52,65 +52,57 @@ int64_t getSizeInBytes(Type type) {
-      data_(static_cast<const char *>(data),
-            static_cast<const char *>(data) + getSizeInBytes(type)) {}
+Tensor::Tensor(ShapedType type) : type_(type) {
+  blob_ = HeapAsmResourceBlob::allocateAndCopy(
```
Done.

sdasgup3 commented on stablehlo/reference/Ops.cpp:
```diff
@@ -28,8 +30,130 @@ Tensor eval(AddOp op, const Tensor &lhs, const Tensor &rhs) {
 }
 
-Tensor eval(ConstantOp op, ElementsAttr value) {
-  return Tensor(value.cast<DenseElementsAttr>());
```
done

sdasgup3 commented on stablehlo/reference/Tensor.h:
```diff
@@ -98,7 +67,17 @@ class Tensor {
+  ArrayRef<char> getData() const { return blob_.getData(); }
+
+  /// Provides access to the underlying mutable storage.
+  MutableArrayRef<char> getMutableData() { return blob_.getMutableData(); }
```
Now these methods should move to the reference counted class `Buffer`.

burmako commented on stablehlo/reference/Ops.cpp:
```diff
@@ -15,6 +15,8 @@ limitations under the License.
 #include "stablehlo/reference/Ops.h"
 
+#include "mlir/IR/AsmState.h"
+#include "mlir/Support/DebugStringHelper.h"
```
We don't need these two includes then?

burmako commented on stablehlo/reference/Tensor.h:
```diff
@@ -46,15 +46,15 @@ class Buffer : public llvm::RefCountedBase<Buffer> {
-  /// Returns the raw data as a byte array.
-  char *getData() { return data_.data(); }
+  /// Provides access to the underlying non-mutable storage.
+  ArrayRef<char> getImmutableData() const { return blob_.getData(); }
```
Any reason to change the name from `getData` to `getImmutableData`? Perhaps let's stick to whatever convention already exists upstream?

burmako commented on stablehlo/reference/Tensor.cpp:
```diff
@@ -339,52 +331,130 @@ void Tensor::dump() const {
-    auto floatValues = llvm::to_vector(
-        llvm::map_range(strData, [&](StringRef strNum) -> APFloat {
-          return APFloat(floatType.getFloatSemantics(), strNum);
+  // Handle floating-point types.
```
Let's leave a comment about why we are using `uint16_t` to store `f16` and `bf16`.

burmako reviewed (commented):
Will leave the rest of the review to Mehdi, including commenting on memory management of blobs.

burmako reviewed (approved):
LGTM from my perspective, still needs a review from Mehdi.

joker-eph commented on docs/interpreter.md:
```diff
@@ -6,10 +6,11 @@
+`mlir::HeapAsmResourceBlob` object representing a mutable blob of tensor
+data laid out as contiguous byte array in
+[major-to-minor order](https://www.tensorflow.org/xla/shapes).
+`detail::Buffer` objects are reference-counted to simplify memory management.
```
Some ownership / lifetime aspect should be documented: a Tensor is an immutable object and copying a tensor doesn't copy the underlying buffer?
(also the Tensor class could be documented, I don't see the header being modified here)

joker-eph commented on stablehlo/reference/Tensor.cpp:
```diff
@@ -71,46 +69,38 @@ Tensor::Tensor() {}
 Tensor::Tensor(ShapedType type)
     : impl_(llvm::makeIntrusiveRefCnt<detail::Buffer>(type)) {}
 
```
If a Tensor is immutable, I'm not sure what do we do with this ctor?

sdasgup3 commented:
Let us port the discussion on mutability of tensor from https://github.com/openxla/stablehlo/pull/101 to this PR.

sdasgup3 commented:
We will be addressing the comments from @joker-eph via [ticket](https://github.com/openxla/stablehlo/issues/233). Planning to merge this as this will unblock writing tests with boolean types and splat.
